From ad18a8b8e9de138c3d89246ac0e25c0467ff5971 Mon Sep 17 00:00:00 2001
From: Nikita Ivanov <nikita.vyach.ivanov@gmail.com>
Date: Fri, 11 Oct 2024 10:50:14 +0200
Subject: [PATCH] Add per client keyboard layout and status bar info

---
 config.def.h |  3 +++
 dwl.c        | 70 +++++++++++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 72 insertions(+), 1 deletion(-)

Index: dwl-v0.7/config.def.h
===================================================================
--- dwl-v0.7.orig/config.def.h
+++ dwl-v0.7/config.def.h
@@ -13,6 +13,9 @@ static const float focuscolor[]
 static const float urgentcolor[]           = COLOR(0xff0000ff);
 /* This conforms to the xdg-protocol. Set the alpha to zero to restore the old behavior */
 static const float fullscreen_bg[]         = {0.1f, 0.1f, 0.1f, 1.0f}; /* You can also use glsl colors */
+/* keyboard layout change notification for status bar */
+static const char  kblayout_file[] = "/tmp/dwl-keymap";
+static const char *kblayout_cmd[]  = {"pkill", "-RTMIN+3", "someblocks", NULL};
 
 /* tagging - TAGCOUNT must be no greater than 31 */
 #define TAGCOUNT (9)
@@ -59,6 +62,9 @@ static const struct xkb_rule_names xkb_r
 	*/
 	.options = NULL,
 };
+/* set this value to correspond with the above */
+static const int xkb_num_layouts = 1;
+static const int xkb_init_layout = 1;
 
 static const int repeat_rate = 25;
 static const int repeat_delay = 600;
@@ -148,6 +154,7 @@ static const Key keys[] = {
 	{ MODKEY,                    XKB_KEY_period,     focusmon,       {.i = WLR_DIRECTION_RIGHT} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_less,       tagmon,         {.i = WLR_DIRECTION_LEFT} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_greater,    tagmon,         {.i = WLR_DIRECTION_RIGHT} },
+	{ MODKEY,                    XKB_KEY_Home,       switchxkbrule,  {-1} },
 	TAGKEYS(          XKB_KEY_1, XKB_KEY_exclam,                     0),
 	TAGKEYS(          XKB_KEY_2, XKB_KEY_at,                         1),
 	TAGKEYS(          XKB_KEY_3, XKB_KEY_numbersign,                 2),
Index: dwl-v0.7/dwl.c
===================================================================
--- dwl-v0.7.orig/dwl.c
+++ dwl-v0.7/dwl.c
@@ -14,6 +14,7 @@
 #include <wayland-server-core.h>
 #include <wlr/backend.h>
 #include <wlr/backend/libinput.h>
+#include <wlr/interfaces/wlr_keyboard.h>
 #include <wlr/render/allocator.h>
 #include <wlr/render/wlr_renderer.h>
 #include <wlr/types/wlr_alpha_modifier_v1.h>
@@ -294,6 +295,8 @@ static void gpureset(struct wl_listener
 static void handlesig(int signo);
 static void incnmaster(const Arg *arg);
 static void inputdevice(struct wl_listener *listener, void *data);
+static void kblayout(KeyboardGroup *kb);
+static void switchxkbrule(const Arg *arg);
 static int keybinding(uint32_t mods, xkb_keysym_t sym);
 static void keypress(struct wl_listener *listener, void *data);
 static void keypressmod(struct wl_listener *listener, void *data);
@@ -413,6 +416,8 @@ static struct wlr_box sgeom;
 static struct wl_list mons;
 static Monitor *selmon;
 
+static unsigned int kblayout_idx = -1;
+
 #ifdef XWAYLAND
 static void activatex11(struct wl_listener *listener, void *data);
 static void associatex11(struct wl_listener *listener, void *data);
@@ -879,6 +884,8 @@ createkeyboard(struct wlr_keyboard *keyb
 
 	/* Add the new keyboard to the group */
 	wlr_keyboard_group_add_keyboard(kb_group->wlr_group, keyboard);
+
+	kblayout(kb_group);
 }
 
 KeyboardGroup *
@@ -1554,6 +1561,68 @@ inputdevice(struct wl_listener *listener
 	wlr_seat_set_capabilities(seat, caps);
 }
 
+void
+kblayout(KeyboardGroup *kb)
+{
+	FILE *f;
+	unsigned int idx = kb->wlr_group->keyboard.modifiers.group;
+
+	// If layout did not change, do nothing
+	if (kblayout_idx == idx)
+		return;
+	kblayout_idx = idx;
+
+	// Save current layout to kblayout_file
+	if (*kblayout_file && (f = fopen(kblayout_file, "w"))) {
+		fputs(xkb_keymap_layout_get_name(kb->wlr_group->keyboard.keymap,
+				idx), f);
+		fclose(f);
+	}
+
+	// Run kblayout_cmd
+	if (kblayout_cmd[0] && fork() == 0) {
+		execvp(kblayout_cmd[0], (char *const *)kblayout_cmd);
+		die("dwl: execvp %s failed:", kblayout_cmd[0]);
+	}
+}
+
+void
+switchxkbrule(const Arg *arg)
+{
+	/*
+	 * Switch keyboard layout/group by index with arg->i.
+	 * If arg->i == -1, then switch to the next layout.
+	 */
+	struct wlr_keyboard *kb = wlr_seat_get_keyboard(seat);
+	struct wlr_keyboard_group *group = kb ? wlr_keyboard_group_from_wlr_keyboard(kb) : NULL;
+
+	/* Copied from wlroots/types/wlr_keyboard_group.c */
+	struct keyboard_group_device {
+		struct wlr_keyboard *keyboard;
+		struct wl_listener key;
+		struct wl_listener modifiers;
+		struct wl_listener keymap;
+		struct wl_listener repeat_info;
+		struct wl_listener destroy;
+		struct wl_list link; // wlr_keyboard_group.devices
+	};
+	struct keyboard_group_device *device;
+	int cur_idx, target_idx;
+
+	if (group) {
+		cur_idx = group->keyboard.modifiers.group;
+		target_idx = (arg->i == -1) ? (++cur_idx % xkb_num_layouts) : arg->i;
+
+		device = wl_container_of(group->devices.next, device, link);
+		wlr_keyboard_notify_modifiers(device->keyboard,
+					      device->keyboard->modifiers.depressed,
+					      device->keyboard->modifiers.latched,
+					      device->keyboard->modifiers.locked,
+					      target_idx
+		);
+	}
+}
+
 int
 keybinding(uint32_t mods, xkb_keysym_t sym)
 {
@@ -1631,6 +1700,8 @@ keypressmod(struct wl_listener *listener
 	/* Send modifiers to the client. */
 	wlr_seat_keyboard_notify_modifiers(seat,
 			&group->wlr_group->keyboard.modifiers);
+
+	kblayout(group);
 }
 
 int
@@ -2249,6 +2320,9 @@ run(char *startup_cmd)
 	if (fd_set_nonblock(STDOUT_FILENO) < 0)
 		close(STDOUT_FILENO);
 
+	/* Set the initial kbd layout */
+	switchxkbrule(&(Arg){ .i = xkb_init_layout });
+
 	printstatus();
 
 	/* At this point the outputs are initialized, choose initial selmon based on
